types/recommendation.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2019 The MayaData Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CStorClusterRecommendation is a kubernetes custom
+// resource that represents recommended configurations
+// for one given input/requirement.
+type CStorClusterRecommendation struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   CStorClusterRecommendationSpec   `json:"spec"`
+	Status CStorClusterRecommendationStatus `json:"status"`
+}
+
+// CStorClusterRecommendationSpec contains recommended
+// block devices topology (blockdevices group by node)
+type CStorClusterRecommendationSpec struct {
+	Topology []DeviceTopology `json:"topology"`
+}
+
+// DeviceTopology contains node identity and identity of
+// the block devices associated with that node
+type DeviceTopology struct {
+	Node         Reference   `json:"node"`
+	BlockDevices []Reference `json:"blockDevices"`
+}
+
+// CStorClusterRecommendationStatus contains key value label pair
+// for nodes and blockdevices. If user select one recommendation
+// nodes and block devices in that recommendation will be labels
+// with NodeLabel and BlockDeviceLabel.
+type CStorClusterRecommendationStatus struct {
```

Comment:
Is Status right term for this structure?

Reply:
Status should be showing the characteristic state of the object. For example, what does the resource under-question has in terms of state. 
For example -- pod can be 'Running' 'pending' etc, pools can be 'online', 'offline'

On a side note -- I am always skeptical in seeing a map in the schema but not denying that it cannot be used.

Reply:
This is the part of schema that is going to be set by the controller only. The controller needs to decide a unique label for nodes & block devices for that instance of time. In other words labels will change if the same request is made again. Since a recommendation can be dynamic i.e. can suggest different every time.

Reply:
This is a map as it directly points to Label & nothing else.
We can think of below as well (as suggested in the design doc) cc @shovanmaity :
```yaml
NodeSelector:
  MatchLabels:
BlockDeviceSelector:
  MatchLabels:
```

Reply:
@AmitKumarDas 
```yaml
spec:
selector:
  nodeSelector:
    matchLabels:
  blockDeviceSelector:
    matchLabels:
status:
```
This looks better :+1:

Reply:
> @sonasingh46 This is a map as it directly points to Label & nothing else.
> We can think of below as well (as suggested in the design doc) cc @shovanmaity :
> 
> ```yaml
> nodeSelector:
>   matchLabels:
> blockDeviceSelector:
>   matchLabels:
> ```
> 
> ```go
> type NodeSelector struct {
>   MatchLabels map[string]string
> }
> type BlockDeviceSelector struct {
>   MatchLabels map[string]string
> }
> ```

Why do we want to introduce one structure for a single map?